repair/repair.cc

```diff
@@ -1989,26 +2024,81 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             }
             auto& strat = erm->get_replication_strategy();
             dht::token_range_vector ranges = strat.get_ranges(myid, *tmptr).get();
-            auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            sstring source_dc_for_keyspace;
+            sstring alt_dc_for_keyspace;
+            auto find_alternative_datacenter = [&] {
+                std::vector<sstring> dcs;
+                dcs.reserve(live_nodes_per_dc.size());
+                std::ranges::copy_if(topology.get_datacenters(), std::back_inserter(dcs), [&] (const auto& dc) {
+                    return dc != *source_dc && !live_nodes_per_dc[dc].empty();
+                });
+                if (!dcs.empty()) {
+                    std::uniform_int_distribution<int> dist(0, dcs.size() - 1);
+                    return dcs[dist(_random_engine)];
+                }
+                return sstring();
+            };
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (source_dc) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    size_t rf = nt_strat.get_replication_factor(*source_dc);
+                    auto lost = lost_nodes_per_dc[*source_dc];
+                    source_dc_for_keyspace = *source_dc;
+
+                    // Look for an alternative source_dc if it's unsafe to repair only
+                    // from the source_dc.
+                    if (rf == 0 || lost > 1 || (lost == 1 && rf == 1)) {
```

Comment:
I am thinking about lost >=1.  

2 ndoes with rf {dc1:3 dc2:3}, 1 node dead in dc1,  we replace the dead node. To protect W=1,R=3, we could use both dc1 and dc2 for replace even if lost = 1. 

Yes, this means replace will always use both local dc and dc with no nodes down (if any). I think it is worth it for safety. Note, we can afford to run full dc repair regularly anyway. It makes less sense to risk safety to use only local dc during replace ops as an optimization to make replace faster.

Reply:
Using another DC is expensive in terms of inter-zone communication, and also slow due to higher latencies so I wouldn't do it by default for replace and lost=1 (meaning we lost only the replaced node in that DC).
If the application is writing with CL=ONE those are the ramifications.
We can add a new `source_dc_for_replace` option to replace, as a follow up, that could specify the source_dc or be set to an empty string or special value like `none` or `false` to specify using all DCs.
If unset, source_dc can be auto-selected by default to be the local DC of the replaced node, as it's done today (after this series).

Reply:
source_dc_for_replace can be a list of dcs for replace ops to choose replica nodes from for simplicity. It could be the local dc itself or remote dc or both. This is simple and flexible. The downside is that it is hard for use to user this option. Users probably not even know they need to choose another dc when they write with CL=ONE and one node is down.  Benny, could you  add some words about the W=1, R=3  and one node down issue in comments. , so we do not forget about this issue. 

FWIW, I am fine to stick with local dc only for replace in case only the replaced node is dead.

Reply:
in v8: removed the `rf=0` special case.  repairing from a different dc won't change anything.  with rf=0, nothing is expected on the replaced/rebuilt node in the keyspace in any case.

when using alt dc, it replaces source_dc, not extending it.

Reply:
> Benny, could you add some words about the W=1, R=3 and one node down issue in comments. , so we do not forget about this issue.

done in v8